### ISSUE TITLE
ObjectId (de)serialization problem

### ISSUE THREAD

**Issue by univerz** [mongodb/bson-rust#264]
`ObjectId` fails to deserialize with `serde_qs` after #220 

```
    use bson::oid::ObjectId;
    let oid1 = ObjectId::from_str("60c2c48f0096ee4d0025cc0c")?;
    let str = serde_qs::to_string(&oid1)?;
    dbg!((&oid1, &str));
    let oid2: ObjectId = serde_qs::from_str(&str)?;
    dbg!(&oid2);
```

* before (`bson = { git = "https://github.com/mongodb/bson-rust", rev = "99b50d0" }`)
```
[allnet/src/main.rs:44] (&oid1, &str) = (
    ObjectId(60c2c48f0096ee4d0025cc0c),
    "%24oid=60c2c48f0096ee4d0025cc0c",
)
[allnet/src/main.rs:46] &oid2 = ObjectId(60c2c48f0096ee4d0025cc0c)
```

* after (`bson = { git = "https://github.com/mongodb/bson-rust", rev = "5f17b3d" }`)
```
[allnet/src/main.rs:44] (&oid1, &str) = (
    ObjectId(60c2c48f0096ee4d0025cc0c),
    "%24oid=60c2c48f0096ee4d0025cc0c",
)
Error: failed with reason: cannot deserialize primitive at the top level.Try deserializing into a struct.
```

(i'm using serde_qs inside thin wrapper to handle forms / query parameters in Rocket framework)

**Comment by agolin95** [comment]
Tracked in: https://jira.mongodb.org/browse/RUST-861

**Comment by isabelatkinson** [comment]
Hi @univerz, thanks for filing this issue! The problem you're encountering is due to the fact that serde_qs only supports structs, maps, and enums at the top level. While our previous implementation of `Deserialize` for `ObjectId` treated an `ObjectId` as a map, the changes in the commit you referenced updated this implementation to deserialize from multiple formats. To fix the error you're encountering, I'd recommend nesting the `ObjectId` in a struct. You can use the `serialize_object_id_as_hex_string` serde helper added in https://github.com/mongodb/bson-rust/commit/b55e958e4de10350b67b7d409ff1160c9bc742cb to get the hex string representation of the `ObjectId` in your query string:

```
#[derive(Serialize, Deserialize)]
struct Oid {
	#[serde(serialize_with = "serialize_object_id_as_hex_string")]
	oid: ObjectId,
}
```

Let us know if you encounter any further issues!

**Comment by univerz** [comment]
thank you :).

> I'd recommend nesting the ObjectId in a struct. 

i considered it, but it's one another small thing for a new team member to remember. however, i'm looking to abandon my custom serde_qs wrapper & use Rocket's `Form`, but they do not support helpers (yet). i'll let you know how that pans out; if not than maybe one subtask for [#595](https://jira.mongodb.org/projects/RUST/issues/RUST-595?filter=allissues) could be to implement `FromFormField` for `ObjectId`.

btw, with changes made a few months ago it's possible to use more composable `serde_with` crate:
```
#[derive(Serialize, Deserialize, Debug, Default)]
pub struct Filter {
    #[serde(default, with = "As::<Option<DisplayFromStr>>")]
    user: Option<ObjectId>,
}
```

**Comment by univerz** [comment]
> To fix the error you're encountering, I'd recommend nesting the ObjectId in a struct.

ah, so it is expected that even [without deserialize_any](https://github.com/mongodb/bson-rust/pull/268/files) `ObjectId -> urlencode -> ObjectId` still fails

```Error: failed with reason: expected ObjectId extended document or hex string```

**Comment by patrickfreed** [comment]
`serde_qs`'s intent is to serialize sets of key-value pairs as a single string. It's only by coincidence that serializing `ObjectId` produces a non-error result, and this is only because it happens to serialize itself as `{ "$oid": <hex string> }` (i.e. something that can be turned into a key value pair). That being said, `serde_qs` is not intended to be used to encode single values like `ObjectId` or `i32` at the top level, instead its meant to encode structs or maps which contain several key value pairs. See the [docs](https://docs.rs/serde_qs/0.8.3/serde_qs/#supported-types) for some more information on that.

So in short, yes it is (kind of) expected behavior that round tripping `ObjectId` itself through `serde_qs` will fail, since it isn't meant for scalar types. However, if you embed the `ObjectId` in a map or a struct, it does safely round trip:
```rust
    let oid = ObjectId::new();
    let c = Config {
        id: oid
    };
    let s = serde_qs::to_string(&c).unwrap();
    let o: Config = serde_qs::from_str(s.as_str()).unwrap();
    println!("{}", s);
    println!("{:?}", o);
```
produces
```
id[%24oid]=60d65ff600d65f6800ccf07c
Config { id: ObjectId("60d65ff600d65f6800ccf07c") }
```

**Comment by univerz** [comment]
thank you both for your time & explanation, i will migrate from my previous "happy accident" to proper solution, ideally rocket's native.

now i feel a litte nervous about initiating potentially unneeded change with performance impact from deserializing through untagged enum.

i did some deserialization (serde_json) benchmarks for 3 bson versions:
```
# last change
bson_enum = { package="bson", git = "https://github.com/mongodb/bson-rust", rev = "a97fb4c" }
# enables deserialization from string
bson_any = { package="bson", git = "https://github.com/mongodb/bson-rust", rev = "5f17b3d" }
# original deserialization from map compatible with my hack
bson_map = { package="bson", git = "https://github.com/mongodb/bson-rust", rev = "99b50d0" }
```

* deserializing `ObjectId` & `struct{ i: ObjectId }` for ObjectId serialized as ObjectId:

```
bson_enum oid from oid  time:   [843.39 ns 850.84 ns 860.18 ns]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

bson_any oid from oid   time:   [525.02 ns 534.26 ns 545.73 ns]
Found 9 outliers among 100 measurements (9.00%)
  4 (4.00%) high mild
  5 (5.00%) high severe

bson_map oid from oid   time:   [521.11 ns 528.78 ns 539.72 ns]
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) high mild
  5 (5.00%) high severe

bson_enum struct from oid
                        time:   [942.16 ns 968.17 ns 1.0140 us]
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe

bson_any struct from oid
                        time:   [550.27 ns 553.01 ns 556.43 ns]
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe

bson_map struct from oid
                        time:   [547.93 ns 560.78 ns 579.10 ns]
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) high mild
  4 (4.00%) high severe

```

* ObjectId serialized as String:

```
bson_enum oid from str  time:   [262.57 ns 264.43 ns 266.44 ns]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

bson_any oid from str   time:   [246.54 ns 250.21 ns 254.72 ns]
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) high mild
  6 (6.00%) high severe

bson_enum struct from str
                        time:   [299.52 ns 302.41 ns 305.94 ns]
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) high mild
  5 (5.00%) high severe

bson_any struct from str
                        time:   [275.42 ns 277.39 ns 280.61 ns]
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe
```


code
```rust
use criterion::{black_box, criterion_group, criterion_main, Criterion};
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize, Debug)]
struct S<O> {
    i: O,
}

fn test_oid(c: &mut Criterion) {
    let oid = bson_map::oid::ObjectId::new();
    let soid = serde_json::to_string(&oid).unwrap();
    let sstruct = serde_json::to_string(&S::<bson_map::oid::ObjectId> { i: oid }).unwrap();

    c.bench_function("bson_enum oid from oid", |b| {
        b.iter(|| {
            let _: bson_enum::oid::ObjectId = serde_json::from_str(black_box(&soid)).unwrap();
        })
    });
    c.bench_function("bson_any oid from oid", |b| {
        b.iter(|| {
            let _: bson_any::oid::ObjectId = serde_json::from_str(black_box(&soid)).unwrap();
        })
    });
    c.bench_function("bson_map oid from oid", |b| {
        b.iter(|| {
            let _: bson_map::oid::ObjectId = serde_json::from_str(black_box(&soid)).unwrap();
        })
    });

    c.bench_function("bson_enum struct from oid", |b| {
        b.iter(|| {
            let _: S<bson_enum::oid::ObjectId> = serde_json::from_str(black_box(&sstruct)).unwrap();
        })
    });
    c.bench_function("bson_any struct from oid", |b| {
        b.iter(|| {
            let _: S<bson_any::oid::ObjectId> = serde_json::from_str(black_box(&sstruct)).unwrap();
        })
    });
    c.bench_function("bson_map struct from oid", |b| {
        b.iter(|| {
            let _: S<bson_map::oid::ObjectId> = serde_json::from_str(black_box(&sstruct)).unwrap();
        })
    });
}

fn test_str(c: &mut Criterion) {
    let oid = bson_map::oid::ObjectId::new();
    let soid = serde_json::to_string(&oid.to_string()).unwrap();
    let sstruct = serde_json::to_string(&S::<String> { i: oid.to_string() }).unwrap();

    c.bench_function("bson_enum oid from str", |b| {
        b.iter(|| {
            let _: bson_enum::oid::ObjectId = serde_json::from_str(black_box(&soid)).unwrap();
        })
    });
    c.bench_function("bson_any oid from str", |b| {
        b.iter(|| {
            let _: bson_any::oid::ObjectId = serde_json::from_str(black_box(&soid)).unwrap();
        })
    });

    c.bench_function("bson_enum struct from str", |b| {
        b.iter(|| {
            let _: S<bson_enum::oid::ObjectId> = serde_json::from_str(black_box(&sstruct)).unwrap();
        })
    });
    c.bench_function("bson_any struct from str", |b| {
        b.iter(|| {
            let _: S<bson_any::oid::ObjectId> = serde_json::from_str(black_box(&sstruct)).unwrap();
        })
    });
}

criterion_group!(benches, test_oid, test_str);
criterion_main!(benches);
```

**Comment by univerz** [comment]
i have noticed new raw deserialization code, so i rerun the benchmark (& removed outlier info to make it more readable ):

deserializing oid from json encoded oid looks better, but worse than before untagged enum:
```
oid from json<oid>/bson_raw
                        time:   [687.92 ns 688.68 ns 689.48 ns]
oid from json<oid>/bson_enum
                        time:   [865.52 ns 866.37 ns 867.29 ns]
oid from json<oid>/bson_any
                        time:   [503.34 ns 503.73 ns 504.13 ns]
oid from json<oid>/bson_map
                        time:   [507.81 ns 508.21 ns 508.63 ns]
```

struct with oid field from json:
```
struct<oid> from json<struct<oid>>/bson_raw
                        time:   [767.13 ns 767.92 ns 768.83 ns]
struct<oid> from json<struct<oid>>/bson_enum
                        time:   [911.56 ns 912.24 ns 912.96 ns]
struct<oid> from json<struct<oid>>/bson_any
                        time:   [531.01 ns 531.65 ns 532.41 ns]
struct<oid> from json<struct<oid>>/bson_map
                        time:   [534.26 ns 534.60 ns 534.98 ns]
```

deserializing oid from json, oid encoded as string is worse:
```
oid from json<str>/bson_raw
                        time:   [255.59 ns 255.88 ns 256.20 ns]
oid from json<str>/bson_enum
                        time:   [249.84 ns 250.12 ns 250.45 ns]
oid from json<str>/bson_any
                        time:   [235.28 ns 235.50 ns 235.73 ns]
```

struct with oid field from json, oid encoded as string:
```
struct<oid> from json<struct<str>>/bson_raw
                        time:   [291.47 ns 291.86 ns 292.26 ns]
struct<oid> from json<struct<str>>/bson_enum
                        time:   [284.82 ns 285.14 ns 285.48 ns]
struct<oid> from json<struct<str>>/bson_any
                        time:   [256.79 ns 257.09 ns 257.40 ns]
```


than i tested deserialization from bytes (== raw bson) - looks better:
```
struct<oid> from bytes<struct<oid>>/bson_raw
                        time:   [773.48 ns 774.75 ns 776.04 ns]
struct<oid> from bytes<struct<oid>>/bson_enum
                        time:   [1.6111 us 1.6126 us 1.6142 us]
struct<oid> from bytes<struct<oid>>/bson_any
                        time:   [1.2543 us 1.2552 us 1.2563 us]
struct<oid> from bytes<struct<oid>>/bson_map
                        time:   [1.2372 us 1.2384 us 1.2397 us]
```


but compared with json (older versions), it looked suspicious. so to get an idea what range to expect, i ran additional benchmarks:

* `ObjectId::from_bytes` == memcopy, just to get an order of magnitude:
```
raw: oid from bytes<oid>
                        time:   [1.7257 ns 1.7272 ns 1.7287 ns]
```

struct with `i64` field (sligtly smaller than ObjectId), good lower bound:
```
raw: struct<i64> from bytes<struct<i64>>
                        time:   [69.259 ns 69.318 ns 69.381 ns]
```

just to be sure that 40x penalty is mostly from struct+serde and not bson, lets see what messagepack can do after a few years of performance tuning:
```
rmp: struct<i64> from bytes<struct<i64>>
                        time:   [15.235 ns 15.248 ns 15.264 ns]
```

struct with `ByteBuf` field filled with ObjectId::bytes (compared with ObjectId there is allocation & unknown length to deal with), good upper bound:
```
raw: struct<ByteBuf> from bytes<struct<oid as ByteBuf>>
                        time:   [98.011 ns 98.110 ns 98.215 ns]
```

motivating example of what msgpack can do with `ByteBuf`:
```
rmp: struct<ByteBuf> from bytes<struct<oid as ByteBuf>>
                        time:   [37.674 ns 37.704 ns 37.737 ns]
```



<details>
<summary>code</summary>

```rust
use criterion::{black_box, criterion_group, criterion_main, Criterion};
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize, Debug)]
struct S<O> {
    i: O,
}

fn test_json_oid(c: &mut Criterion) {
    let oid = bson_map::oid::ObjectId::new();
    let soid = serde_json::to_string(&oid).unwrap();
    let sstruct = serde_json::to_string(&S::<bson_map::oid::ObjectId> { i: oid }).unwrap();

    let mut group = c.benchmark_group("oid from json<oid>");
    group.bench_function("bson_raw", |b| {
        b.iter(|| {
            let _: bson_raw::oid::ObjectId = serde_json::from_str(black_box(&soid)).unwrap();
        })
    });
    group.bench_function("bson_enum", |b| {
        b.iter(|| {
            let _: bson_enum::oid::ObjectId = serde_json::from_str(black_box(&soid)).unwrap();
        })
    });
    group.bench_function("bson_any", |b| {
        b.iter(|| {
            let _: bson_any::oid::ObjectId = serde_json::from_str(black_box(&soid)).unwrap();
        })
    });
    group.bench_function("bson_map", |b| {
        b.iter(|| {
            let _: bson_map::oid::ObjectId = serde_json::from_str(black_box(&soid)).unwrap();
        })
    });
    group.finish();

    let mut group = c.benchmark_group("struct<oid> from json<struct<oid>>");
    group.bench_function("bson_raw", |b| {
        b.iter(|| {
            let _: S<bson_raw::oid::ObjectId> = serde_json::from_str(black_box(&sstruct)).unwrap();
        })
    });
    group.bench_function("bson_enum", |b| {
        b.iter(|| {
            let _: S<bson_enum::oid::ObjectId> = serde_json::from_str(black_box(&sstruct)).unwrap();
        })
    });
    group.bench_function("bson_any", |b| {
        b.iter(|| {
            let _: S<bson_any::oid::ObjectId> = serde_json::from_str(black_box(&sstruct)).unwrap();
        })
    });
    group.bench_function("bson_map", |b| {
        b.iter(|| {
            let _: S<bson_map::oid::ObjectId> = serde_json::from_str(black_box(&sstruct)).unwrap();
        })
    });
    group.finish();
}

fn test_json_str(c: &mut Criterion) {
    let oid = bson_map::oid::ObjectId::new();
    let soid = serde_json::to_string(&oid.to_string()).unwrap();
    let sstruct = serde_json::to_string(&S::<String> { i: oid.to_string() }).unwrap();

    let mut group = c.benchmark_group("oid from json<str>");
    group.bench_function("bson_raw", |b| {
        b.iter(|| {
            let _: bson_raw::oid::ObjectId = serde_json::from_str(black_box(&soid)).unwrap();
        })
    });
    group.bench_function("bson_enum", |b| {
        b.iter(|| {
            let _: bson_enum::oid::ObjectId = serde_json::from_str(black_box(&soid)).unwrap();
        })
    });
    group.bench_function("bson_any", |b| {
        b.iter(|| {
            let _: bson_any::oid::ObjectId = serde_json::from_str(black_box(&soid)).unwrap();
        })
    });
    group.finish();

    let mut group = c.benchmark_group("struct<oid> from json<struct<str>>");
    group.bench_function("bson_raw", |b| {
        b.iter(|| {
            let _: S<bson_raw::oid::ObjectId> = serde_json::from_str(black_box(&sstruct)).unwrap();
        })
    });
    group.bench_function("bson_enum", |b| {
        b.iter(|| {
            let _: S<bson_enum::oid::ObjectId> = serde_json::from_str(black_box(&sstruct)).unwrap();
        })
    });
    group.bench_function("bson_any", |b| {
        b.iter(|| {
            let _: S<bson_any::oid::ObjectId> = serde_json::from_str(black_box(&sstruct)).unwrap();
        })
    });
    group.finish();
}

fn test_bytes_oid(c: &mut Criterion) {
    let oid = bson_map::oid::ObjectId::new();
    let mut vstruct: Vec<u8> = Vec::new();
    bson_map::to_document(&S::<bson_map::oid::ObjectId> { i: oid.clone() }).unwrap().to_writer(&mut vstruct).unwrap();

    let mut group = c.benchmark_group("struct<oid> from bytes<struct<oid>>");
    group.bench_function("bson_raw", |b| {
        b.iter(|| {
            let _: S<bson_raw::oid::ObjectId> = bson_raw::from_slice(black_box(vstruct.as_slice())).unwrap();
        })
    });
    group.bench_function("bson_enum", |b| {
        b.iter(|| {
            let _: S<bson_enum::oid::ObjectId> =
                bson_enum::from_document(bson_enum::Document::from_reader(black_box(&mut vstruct.as_slice())).unwrap())
                    .unwrap();
        })
    });
    group.bench_function("bson_any", |b| {
        b.iter(|| {
            let _: S<bson_any::oid::ObjectId> =
                bson_any::from_document(bson_any::Document::from_reader(black_box(&mut vstruct.as_slice())).unwrap())
                    .unwrap();
        })
    });
    group.bench_function("bson_map", |b| {
        b.iter(|| {
            let _: S<bson_map::oid::ObjectId> =
                bson_map::from_document(bson_map::Document::from_reader(black_box(&mut vstruct.as_slice())).unwrap())
                    .unwrap();
        })
    });
    group.finish();
}

fn test_bytes_mixed(c: &mut Criterion) {
    let oid = bson_map::oid::ObjectId::new();

    let bytes = oid.bytes();
    c.bench_function("raw: oid from bytes<oid>", |b| {
        b.iter(|| {
            let _ = bson_raw::oid::ObjectId::from_bytes(black_box(bytes));
        })
    });

    let mut vstruct: Vec<u8> = Vec::new();
    bson_map::to_document(&S::<i64> { i: i64::MAX }).unwrap().to_writer(&mut vstruct).unwrap();
    c.bench_function("raw: struct<i64> from bytes<struct<i64>>", |b| {
        b.iter(|| {
            let _: S<i64> = bson_raw::from_slice(black_box(vstruct.as_slice())).unwrap();
        })
    });

    let vstruct = rmp_serde::to_vec(&S::<i64> { i: i64::MAX }).unwrap();
    c.bench_function("rmp: struct<i64> from bytes<struct<i64>>", |b| {
        b.iter(|| {
            let _: S<i64> = rmp_serde::from_slice(vstruct.as_slice()).unwrap();
        })
    });

    use serde_bytes::ByteBuf;
    let mut vstruct: Vec<u8> = Vec::new();
    bson_map::to_document(&S::<ByteBuf> { i: ByteBuf::from(oid.bytes()) }).unwrap().to_writer(&mut vstruct).unwrap();
    c.bench_function("raw: struct<ByteBuf> from bytes<struct<oid as ByteBuf>>", |b| {
        b.iter(|| {
            let _: S<ByteBuf> = bson_raw::from_slice(black_box(vstruct.as_slice())).unwrap();
        })
    });

    let vstruct = rmp_serde::to_vec(&S::<ByteBuf> { i: ByteBuf::from(oid.bytes()) }).unwrap();
    c.bench_function("rmp: struct<ByteBuf> from bytes<struct<oid as ByteBuf>>", |b| {
        b.iter(|| {
            let _: S<ByteBuf> = rmp_serde::from_read_ref(&vstruct).unwrap();
        })
    });
}

criterion_group!(benches, test_json_oid, test_json_str, test_bytes_oid, test_bytes_mixed);
criterion_main!(benches);
```

</details>

**Comment by univerz** [comment]
TLDR (for raw deserialization): `struct {i: ObjectId}` (base "class" of most of the mongodb documents) takes 770ns to deserialize, but 69-98ns is expected & possibly 15-37ns with further tuning as msgpack is close to mongodb format.

**Comment by univerz** [comment]
i forget to add blackboxing to msgpack, but numbers holds.

**Comment by patrickfreed** [comment]
Thanks so much for the detailed benchmarks @univerz! 

A few observations:
- Deserializing an actual `ObjectId` from raw BSON will always be slower than deserializing a regular `ByteBuf`, as we have to go through the extended JSON format instead of directly through serde's data model.
- We _should_ be able to meet msgpack's performance with stuff that are in serde's data model though, such as in the `struct<i64>` case. I filed RUST-888 to cover the work for matching `rmp_serde` here.
- The main reason deserializing `ObjectId` is slower in raw BSON is that we currently have to allocate a string and then parse it, even though wer're starting in raw BSON and ending in raw BSON. If we introduced a shortcut here to go straight from bytes to bytes, we could get close to matching our `ByteBuf` performance. I filed RUST-889 to track this work.
- Optimizing this could net ~500ns per `ObjectId`, which is still only 0.0005ms and, in the face of network latency and other DB processing, largely negligible. For non-driver workloads this could be a not-insignificant factor, however.

Thanks again for diving so deep into this!

**Comment by univerz** [comment]
> in the face of network latency and other DB processing, largely negligible

i agree on latency, but this is limiting througput too. when i was thinking about how much is the original 1.6us value bad, i realized that deserialization time of just ObjectIds from all documents (medium size) i'm using to render an webpage is rougly the same as rendering html template with all that data & there are formatted chrono::DateTime's (timezones) & nontrivial state in a lot of generated urls (including `ObjectId::to_hex()`). i'm really looking forward to finally consolidate 3different mongo drivers/versions that i use across projects because of really bad perf.

i managed to at least skip bson  (i do not know serde enough to hack bytes -> bytes & manage full compatibility with extjson), 774ns -> 487ns
```diff --git a/src/de/serde.rs b/src/de/serde.rs
index 7831a90..f5f6afe 100644
--- a/src/de/serde.rs
+++ b/src/de/serde.rs
@@ -40,24 +40,18 @@ impl<'de> Deserialize<'de> for ObjectId {
         #[derive(serde::Deserialize)]
         #[serde(untagged)]
         enum OidHelper {
+            OidS {
+                #[serde(rename = "$oid")]
+                s: String,
+            },
             HexString(String),
-            Bson(Bson),
         }
 
         match OidHelper::deserialize(deserializer)
             .map_err(|_| de::Error::custom("expected ObjectId extended document or hex string"))?
         {
+            OidHelper::OidS { s } => ObjectId::parse_str(s).map_err(de::Error::custom),
             OidHelper::HexString(s) => ObjectId::parse_str(&s).map_err(de::Error::custom),
-            OidHelper::Bson(bson) => match bson {
-                Bson::ObjectId(oid) => Ok(oid),
-                bson => {
-                    let err = format!(
-                        "expected objectId extended document or hex string, found {}",
-                        bson
-                    );
-                    Err(de::Error::invalid_type(Unexpected::Map, &&err[..]))
-                }
-            },
         }
     }
 }
```

**Comment by univerz** [comment]
managed to hack it with `deserializer.is_human_readable()` ==> without additional changes it breaks if user tries to (de)serialize ObjectId from another binary format, but perf is as expected - 72ns.

```
oid from json<oid>/bson_raw                                                                            
                        time:   [312.61 ns 312.90 ns 313.22 ns]
struct<oid> from json<struct<oid>>/bson_raw                                                                            
                        time:   [359.77 ns 360.08 ns 360.41 ns]
oid from json<str>/bson_raw                                                                            
                        time:   [218.09 ns 218.32 ns 218.59 ns]
struct<oid> from json<struct<str>>/bson_raw                                                                            
                        time:   [244.00 ns 244.24 ns 244.51 ns]
struct<oid> from bytes<struct<oid>>/bson_raw                                                                            
                        time:   [71.898 ns 71.955 ns 72.018 ns]
```

<details>
<summary>diff</summary>

```diff --git a/src/de/raw.rs b/src/de/raw.rs
index 9927662..1086f0e 100644
--- a/src/de/raw.rs
+++ b/src/de/raw.rs
@@ -241,6 +241,42 @@ impl<'de, 'a> serde::de::Deserializer<'de> for &'a mut Deserializer<'de> {
         }
     }
 
+    fn deserialize_bytes<V>(self, visitor: V) -> Result<V::Value>
+    where
+        V: serde::de::Visitor<'de>,
+    {
+        match self.current_type {
+            ElementType::ObjectId => visitor.visit_borrowed_bytes(self.bytes.read_slice(12)?),
+            ElementType::Binary => {
+                let len = read_i32(&mut self.bytes)?;
+                if !(0..=MAX_BSON_SIZE).contains(&len) {
+                    return Err(Error::invalid_length(
+                        len as usize,
+                        &format!("binary length must be between 0 and {}", MAX_BSON_SIZE).as_str(),
+                    ));
+                }
+                let subtype = BinarySubtype::from(read_u8(&mut self.bytes)?);
+                match subtype {
+                    BinarySubtype::Generic => {
+                        visitor.visit_borrowed_bytes(self.bytes.read_slice(len as usize)?)
+                    }
+                    _ => {
+                        let binary = Binary::from_reader_with_len_and_payload(
+                            &mut self.bytes,
+                            len,
+                            subtype,
+                        )?;
+                        let mut d = BinaryDeserializer::new(binary);
+                        visitor.visit_map(BinaryAccess {
+                            deserializer: &mut d,
+                        })
+                    }
+                }
+            }
+            _ => self.deserialize_any(visitor),
+        }
+    }
+
     #[inline]
     fn deserialize_option<V>(self, visitor: V) -> Result<V::Value>
     where
@@ -275,7 +311,7 @@ impl<'de, 'a> serde::de::Deserializer<'de> for &'a mut Deserializer<'de> {
     }
 
     forward_to_deserialize_any! {
-        bool char str bytes byte_buf unit unit_struct string
+        bool char str byte_buf unit unit_struct string
         identifier newtype_struct seq tuple tuple_struct struct
         map ignored_any i8 i16 i32 i64 u8 u16 u32 u64 f32 f64
     }
diff --git a/src/de/serde.rs b/src/de/serde.rs
index 7831a90..4d384dd 100644
--- a/src/de/serde.rs
+++ b/src/de/serde.rs
@@ -32,32 +32,72 @@ use super::raw::Decimal128Access;
 
 pub(crate) struct BsonVisitor;
 
+pub(crate) struct ObjectIdVisitor;
+impl<'de> Visitor<'de> for ObjectIdVisitor {
+    type Value = ObjectId;
+
+    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("ObjectId (24-character hex || 12bytes || extjson)")
+    }
+
+    #[inline]
+    fn visit_str<E: de::Error>(self, value: &str) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        ObjectId::parse_str(value).map_err(|_| {
+            E::invalid_value(
+                Unexpected::Str(value),
+                &"24-character, big-endian hex string",
+            )
+        })
+    }
+
+    #[inline]
+    fn visit_string<E: de::Error>(self, value: String) -> Result<Self::Value, E> {
+        ObjectId::parse_str(value.as_str()).map_err(|_| {
+            E::invalid_value(
+                Unexpected::Str(&value),
+                &"24-character, big-endian hex string",
+            )
+        })
+    }
+
+    #[inline]
+    fn visit_borrowed_bytes<E>(self, value: &'de [u8]) -> Result<Self::Value, E>
+    where
+        E: Error,
+    {
+        ObjectId::from_reader(value)
+            .map_err(|_| E::invalid_value(Unexpected::Bytes(value), &"12 bytes"))
+    }
+
+    #[inline]
+    fn visit_map<V>(self, mut visitor: V) -> Result<Self::Value, V::Error>
+    where
+        V: MapAccess<'de>,
+    {
+        match BsonVisitor.visit_map(&mut visitor)? {
+            Bson::ObjectId(oid) => Ok(oid),
+            bson => {
+                let err = format!(
+                    "expected objectId extended document or hex string, found {}",
+                    bson
+                );
+                Err(de::Error::invalid_type(Unexpected::Map, &&err[..]))
+            }
+        }
+    }
+}
 impl<'de> Deserialize<'de> for ObjectId {
     fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
     where
         D: de::Deserializer<'de>,
     {
-        #[derive(serde::Deserialize)]
-        #[serde(untagged)]
-        enum OidHelper {
-            HexString(String),
-            Bson(Bson),
-        }
-
-        match OidHelper::deserialize(deserializer)
-            .map_err(|_| de::Error::custom("expected ObjectId extended document or hex string"))?
-        {
-            OidHelper::HexString(s) => ObjectId::parse_str(&s).map_err(de::Error::custom),
-            OidHelper::Bson(bson) => match bson {
-                Bson::ObjectId(oid) => Ok(oid),
-                bson => {
-                    let err = format!(
-                        "expected objectId extended document or hex string, found {}",
-                        bson
-                    );
-                    Err(de::Error::invalid_type(Unexpected::Map, &&err[..]))
-                }
-            },
+        if !deserializer.is_human_readable() {
+            deserializer.deserialize_bytes(ObjectIdVisitor)
+        } else {
+            deserializer.deserialize_any(ObjectIdVisitor)
         }
     }
 }
```

</details>



\+ it should be possible to use something similar for ObjectIdDeserializer, because deserialization of (maybe not only, didn't test) bytes<struct<oid>> to **`Document` regressed** compared to previous versions (624ns vs 260ns).

**Comment by univerz** [comment]
hm, even serializing `S<ObjectId>` to msgpack and deserialization back to struct & `rmp::Value` works, so maybe it is not as horrible hack as i expected. it looks like the only change is that if user wants to deserialize ObjectId to bytes, it would work (objectid is defined as 12bytes, so it's ok)

**Comment by patrickfreed** [comment]
Your solution was included in `2.0.0-beta.3`, see #289. Thanks so much for helping out with this! 

Going to close this out now; to track the remaining work needed to match `rmp_serde`, please follow RUST-888.
